## 1. The World Bank's international debt data
In this notebook, we will explore the topic of countries taking on debt to manage their economies. It is not just individuals who take on debt to manage their necessities; countries also borrow money to fund infrastructure spending and other important expenses. The [World Bank](https://www.worldbank.org/en/home) is one organization that provides debt to countries for this purpose.

We will be analyzing international debt data collected by The World Bank in this notebook. This dataset contains information about the amount of debt (in USD) owed by developing countries across various categories. Through our analysis, we aim to answer questions such as:

- What is the total amount of debt owed by the countries listed in the dataset?
- Which country has the highest amount of debt, and what is that amount?
- What is the average amount of debt owed by countries across different debt indicators?

In [1]:
import os
import psycopg2
import pandas as pd




In [2]:
#pip install ipython-sql(this enables the use of SQL magic functions that contain % and %% , allowing you to write SQL style code right)
#pip install sqlalchemy(it will mainly be used to store SQL queries into a pandas dataframe.)
#pip install psycopg2 (forPostgreSQL, you would use psycopg2:)

# Load the IPython SQL extension.
%load_ext sql

In [3]:
from sqlalchemy import create_engine, Column, Integer, String, Numeric
from sqlalchemy.orm import declarative_base
from sqlalchemy.pool import NullPool

# Create a database engine that connects to a PostgreSQL database.
# Retrieve database credentials from environment variables
db_username = os.environ.get('DB_USERNAME')
db_password = os.environ.get('DB_PASSWORD')
db_hostname = os.environ.get('DB_HOSTNAME')
engine = create_engine(f'postgresql://{db_username}:{db_password}@{db_hostname}/postgres')

# Create a base class for declarative class definitions.
Base = declarative_base()


# Define a new class for a table in the database.
class MyTable(Base):
    __tablename__ = 'my_table'

    id = Column(Integer, primary_key=True)
    country_name = Column(String(50))
    country_code = Column(String(50))
    indicator_name = Column(String)
    indicator_code = Column(String)
    debt = Column(Numeric)

Base.metadata.create_all(engine)



In [4]:
# Specify the path to the CSV file to be inserted into the database.
csv_file_path = 'G:/My Drive/Analytics/international_debt.csv'

# Read the CSV file into a pandas DataFrame.
df = pd.read_csv(csv_file_path)

# Use the `to_sql()` method of the DataFrame object to insert the data into the database table.
df.to_sql('my_table', engine, if_exists='append', index=False)




357

In [5]:
%sql postgresql://{db_username}:{db_password}@{db_hostname}/postgres

In [6]:
%%sql 
SELECT distinct(country_name)FROM my_table


 * postgresql://postgres:***@localhost/postgres
124 rows affected.


[('Indonesia',),
 ('Bangladesh',),
 ('Iran, Islamic Rep.',),
 ('Cameroon',),
 ('St. Lucia',),
 ('Uganda',),
 ('Montenegro',),
 ('Macedonia, FYR',),
 ('Jordan',),
 ('Syrian Arab Republic',),
 ('Dominican Republic',),
 ('Cambodia',),
 ('Papua New Guinea',),
 ('Sri Lanka',),
 ('Uzbekistan',),
 ('Colombia',),
 ('Albania',),
 ('Ukraine',),
 ('Grenada',),
 ('Sao Tome and Principe',),
 ("Cote d'Ivoire",),
 ('Algeria',),
 ('Maldives',),
 ('Djibouti',),
 ('Senegal',),
 ('Kenya',),
 ('Ghana',),
 ('Zambia',),
 ('Madagascar',),
 ('Sierra Leone',),
 ('Liberia',),
 ('Bosnia and Herzegovina',),
 ('Philippines',),
 ('Benin',),
 ('Cabo Verde',),
 ('Guinea',),
 ('Turkey',),
 ('Nigeria',),
 ('Rwanda',),
 ('Zimbabwe',),
 ('Tajikistan',),
 ('Congo, Rep.',),
 ('Comoros',),
 ('China',),
 ('Belarus',),
 ('Armenia',),
 ('Lesotho',),
 ('Paraguay',),
 ('Gabon',),
 ('Dominica',),
 ('Serbia',),
 ('Mauritius',),
 ('Angola',),
 ('Gambia, The',),
 ('Vanuatu',),
 ('Georgia',),
 ('Malawi',),
 ('Burundi',),
 ('Bhutan',),
 ('Solomon Islands',),
 ('Thailand',),
 ('Togo',),
 ('Venezuela, RB',),
 ('Burkina Faso',),
 ('El Salvador',),
 ('Eswatini',),
 ('Fiji',),
 ('Congo, Dem. Rep.',),
 ('Eritrea',),
 ('Egypt, Arab Rep.',),
 ('Azerbaijan',),
 ('Afghanistan',),
 ('India',),
 ('Chad',),
 ('Timor-Leste',),
 ('Vietnam',),
 ('Kyrgyz Republic',),
 ('South Africa',),
 ('Peru',),
 ('Kazakhstan',),
 ('Central African Republic',),
 ('Kosovo',),
 ('Jamaica',),
 ('Yemen, Rep.',),
 ('Mongolia',),
 ('Mauritania',),
 ('Mozambique',),
 ('IDA only',),
 ('Ecuador',),
 ('Honduras',),
 ('Belize',),
 ('Botswana',),
 ('Russian Federation',),
 ('Pakistan',),
 ('Romania',),
 ('Brazil',),
 ('Guatemala',),
 ('Guinea-Bissau',),
 ('Bolivia',),
 ('Ethiopia',),
 ('Niger',),
 ('Bulgaria',),
 ('Lao PDR',),
 ('Turkmenistan',),
 ('South Asia',),
 ('Tunisia',),
 ('Sudan',),
 ('Mali',),
 ('Morocco',),
 ('Moldova',),
 ('Myanmar',),
 ('St. Vincent and the Grenadines',),
 ('Nicaragua',),
 ('Mexico',),
 ('Tonga',),
 ('Nepal',),
 ('Guyana',),
 ('Tanzania',),
 ('Lebanon',),
 ('Costa Rica',),
 ('Least developed countries: UN classification',),
 ('Haiti',),
 ('Samoa',),
 ('Somalia',)]

**If you’d like to store your query in a pandas DataFrame, this is where sqlalchemy comes in. Create a dataframe object using the command pd.read_sql() . It takes two arguments:**

In [7]:
result = %sql SELECT * FROM my_table LIMIT 3
df = result.DataFrame()

print(df)


 * postgresql://postgres:***@localhost/postgres
3 rows affected.
   id country_name country_code  \
0   1  Afghanistan          AFG   
1   2  Afghanistan          AFG   
2   3  Afghanistan          AFG   

                                      indicator_name  indicator_code  \
0  Disbursements on external debt, long-term (DIS...  DT.DIS.DLXF.CD   
1  Interest payments on external debt, long-term ...  DT.INT.DLXF.CD   
2                  PPG, bilateral (AMT, current US$)  DT.AMT.BLAT.CD   

         debt  
0  72894453.7  
1  53239440.1  
2  61739336.9  



## Finding the number of distinct countries
In the first ten rows, we can observe the amount of debt owed by Afghanistan across various debt indicators. However, it's unclear how many countries are represented in the table due to repetitions in country names, as a country may have debt in multiple debt indicators.

Without knowing the count of unique countries, our statistical analyses cannot be conducted comprehensively. Therefore, this section aims to determine the number of distinct countries in the table.





In [8]:
%%sql
SELECT 
    COUNT(DISTINCT country_name) AS total_distinct_countries
FROM my_table;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(124,)]

## Finding out the distinct debt indicators
The table contains a total of 124 countries, as we can observe. In the previous section, we noted the presence of the "indicator_name" column, which provides a brief description of the purpose for which the debt was incurred. Adjacent to this column is the "indicator_code" column, which signifies the category to which the debts belong. Understanding the different debt indicators can provide insight into the potential areas for which a country may have incurred debt.


In [9]:
%%sql
SELECT count(DISTINCT indicator_code) AS distinct_debt_indicators
FROM my_table
ORDER BY distinct_debt_indicators

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[(25,)]

%%sql
SELECT DISTINCT indicator_code AS distinct_debt_indicators
FROM my_table
ORDER BY distinct_debt_indicators

## Total amount of debt owed by each countries

In [10]:
%%sql
SELECT 
    country_name, 
    SUM(debt) AS total_debt
FROM my_table
GROUP BY country_name
ORDER BY total_debt DESC
LIMIT 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[('China', Decimal('16576022694624.2')),
 ('Brazil', Decimal('16276190036201.2')),
 ('South Asia', Decimal('14361305991489.6')),
 ('Least developed countries: UN classification', Decimal('12347097582005.6')),
 ('Russian Federation', Decimal('11094765321051.0')),
 ('IDA only', Decimal('10384791378093.0')),
 ('Turkey', Decimal('8765293966093.8')),
 ('India', Decimal('7750369535598.8')),
 ('Mexico', Decimal('7226613600644.0')),
 ('Indonesia', Decimal('6579270408321.6'))]

## Average amount of debt across indicators
China was identified as the debtor country. For a detailed breakdown of China's debts, please refer to this [link.](https://datatopics.worldbank.org/debt/ids/country/CHN)

We have gained a basic understanding of the dataset and its summary statistics. Additionally, we have familiarized ourselves with the various debt indicators for which countries owe their debts. To further our analysis, we can explore the average debt amount owed by each country. This information can provide insights into the distribution of debt amounts across different indicators.

In [11]:
%%sql
SELECT 
    indicator_code AS debt_indicator,
    indicator_name,
    AVG(debt) AS average_debt
FROM my_table
GROUP BY debt_indicator, indicator_name
ORDER BY average_debt DESC
LIMIT 10;

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[('DT.AMT.DLXF.CD', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('5904868401.52661290')),
 ('DT.AMT.DPNG.CD', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', Decimal('5161194333.82531646')),
 ('DT.DIS.DLXF.CD', 'Disbursements on external debt, long-term (DIS, current US$)', Decimal('2152041216.89756098')),
 ('DT.DIS.OFFT.CD', 'PPG, official creditors (DIS, current US$)', Decimal('1958983452.88606557')),
 ('DT.AMT.PRVT.CD', 'PPG, private creditors (AMT, current US$)', Decimal('1803694101.97040816')),
 ('DT.INT.DLXF.CD', 'Interest payments on external debt, long-term (INT, current US$)', Decimal('1644024067.63145161')),
 ('DT.DIS.BLAT.CD', 'PPG, bilateral (DIS, current US$)', Decimal('1223139290.39646018')),
 ('DT.INT.DPNG.CD', 'Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)', Decimal('1220410844.42658228')),
 ('DT.AMT.OFFT.CD', 'PPG, official creditors (AMT, current US$)', Decimal('1191187963.09758065')),
 ('DT.AMT.PBND.CD', 'PPG, bonds (AMT, current US$)', Decimal('1082623947.66086957'))]

## The highest amount of principal repayments
The debt indicator DT.AMT.DLXF.CD ranks highest in terms of average debt amount. This category pertains to long-term debt repayment, which countries incur to obtain immediate capital. For more information on this debt category, please refer to this [source.]

Notably, there is a substantial difference in debt amounts between the top two indicators and the rest. This suggests that the first two indicators may represent the most severe debt categories for countries.

To investigate this further, we can identify the country with the highest long-term debt (DT.AMT.DLXF.CD) and gain insights into its economic condition. Since economic disruptions may vary across countries, this finding can provide more specific information about that particular country's situation.

In [12]:
%%sql
SELECT 
    country_name, 
    indicator_name,
    debt
FROM my_table
WHERE debt = (SELECT 
                 MAX(debt)
             FROM my_table
             WHERE indicator_code = 'DT.AMT.DLXF.CD')
LIMIT 1;

 * postgresql://postgres:***@localhost/postgres
1 rows affected.


[('China', 'Principal repayments on external debt, long-term (AMT, current US$)', Decimal('96218620836.0'))]

## The most common debt indicator
According to The World Bank, China has the highest amount of long-term debt (DT.AMT.DLXF.CD) among all countries. It is always a good practice to verify our analyses to ensure their accuracy and reliability.

While long-term debt (DT.AMT.DLXF.CD) has the highest average debt amount, it remains unclear if it is the most frequent debt category among countries. To investigate this further, let's explore which debt indicators are most commonly used by countries to incur debt.






In [13]:
%%sql
SELECT indicator_code, COUNT(indicator_code) AS indicator_count
FROM my_table
GROUP BY indicator_code
ORDER BY indicator_count DESC, indicator_code DESC
LIMIT 20

 * postgresql://postgres:***@localhost/postgres
20 rows affected.


[('DT.INT.OFFT.CD', 7192),
 ('DT.INT.MLAT.CD', 7192),
 ('DT.INT.DLXF.CD', 7192),
 ('DT.AMT.OFFT.CD', 7192),
 ('DT.AMT.MLAT.CD', 7192),
 ('DT.AMT.DLXF.CD', 7192),
 ('DT.DIS.DLXF.CD', 7134),
 ('DT.INT.BLAT.CD', 7076),
 ('DT.DIS.OFFT.CD', 7076),
 ('DT.AMT.BLAT.CD', 7076),
 ('DT.DIS.MLAT.CD', 6960),
 ('DT.DIS.BLAT.CD', 6554),
 ('DT.INT.PRVT.CD', 5684),
 ('DT.AMT.PRVT.CD', 5684),
 ('DT.INT.PCBK.CD', 4872),
 ('DT.AMT.PCBK.CD', 4872),
 ('DT.INT.DPNG.CD', 4582),
 ('DT.AMT.DPNG.CD', 4582),
 ('DT.INT.PBND.CD', 4002),
 ('DT.AMT.PBND.CD', 4002)]

## Other viable debt issues and conclusion
Our dataset contains a total of six debt indicators, including the DT.AMT.DLXF.CD category. This suggests that all listed countries may be experiencing a similar economic issue. However, this is just a part of the overall picture.

Let's shift our focus back to the amount of debt and explore the maximum debt amount owed by each country. This information can reveal other potential economic issues that a country might be facing.

In summary, this notebook provides an overview of the debt owed by countries worldwide, including summary statistics and interesting insights. We have also verified our findings to ensure their accuracy and reliability.

In [14]:
%%sql
SELECT country_name,
MAX(debt) AS maximum_debt
FROM my_table
GROUP BY country_name
ORDER BY maximum_debt DESC
LIMIT 10

 * postgresql://postgres:***@localhost/postgres
10 rows affected.


[('China', Decimal('96218620836.0')),
 ('Brazil', Decimal('90041840304.0')),
 ('Russian Federation', Decimal('66589761834.0')),
 ('Turkey', Decimal('51555031006.0')),
 ('South Asia', Decimal('48756295898.0')),
 ('Least developed countries: UN classification', Decimal('40160766262.0')),
 ('IDA only', Decimal('34531188113.0')),
 ('India', Decimal('31923507001.0')),
 ('Indonesia', Decimal('30916112654.0')),
 ('Kazakhstan', Decimal('27482093686.0'))]